core/trino-spi/src/main/java/io/trino/spi/predicate/Utils.java

```diff
@@ -64,4 +72,48 @@ static RuntimeException handleThrowable(Throwable throwable)
         }
         return new RuntimeException(throwable);
     }
+
+    /**
+     * Try to expand {@code valueSet} into a discrete set of (at most {@code limit}) objects.
+     * For example: [1, 5] can be expanded into {1, 2, 3, 4, 5}.
+     * If the data type is not supported or the expansion results in too many values, {@code Optional.empty()} is returned.
+     */
+    public static Optional<Collection<Object>> tryExpandRanges(ValueSet valuesSet, int limit)
+    {
+        if (!(valuesSet instanceof SortedRangeSet) || limit <= 0) {
+            return Optional.empty();
+        }
+        List<Range> ranges = valuesSet.getRanges().getOrderedRanges();
+        Type type = valuesSet.getType();
+        List<Object> result = new ArrayList<>(limit);
+        for (Range range : ranges) {
+            if (range.isLowUnbounded() || range.isHighUnbounded()) {
+                return Optional.empty();
+            }
+            if (range.isSingleValue()) {
```

Comment:
nit: you can move this if block to beginning of the loop since `isSingleValue` is true only when `lowValue.isPresent() && highValue.isPresent()`

Reply:
The `isSingleValue` case is a subset of the normal multi-range case. Why not let the code below (type.getDiscreteValues, etc) handle it and simplify this block?

Reply:
@raunaqmorarka WDYT?

Reply:
I tent to agree with @raunaqmorarka . unless we measure it, special-casing `range.isSingleValue()` feels reasonable to me

Reply:
I would argue the opposite. Unless we can prove it matters, the simpler code is preferable.

Reply:
Changing to simpler code :)

Reply:
@martint while i agree that simpler code is preferable is a good general rule, we've had enough cases around TupleDomain handling to grow some intuition about performance requirements of the code, and what kind of changes make the code more performant. While intuition can be obviously deceptive, unless we benchmark it, we have "keep it more performant intuition" vs "keep it more readable intuition".
Does it mean we need to measure it?

Reply:
The single value case is the common case in domains generated for dynamic filters. Dynamic filter collection can produce a domain which consists of many single values and a small number of min-max ranges.
Special casing for single value also gives the nice property that if `tryExpandRanges` is called on a discrete set, then it will give back the discrete set (if it is under valuesLimit) even if `Type#getDiscreteValues` has not been implemented for the given type.

Reply:
@raunaqmorarka let's improve this as a follow-up